### ISSUE TITLE
'host' setting does not change binding of console ports

### ISSUE THREAD

**Issue by candlerb** [GNS3/gns3-server#1667]
(gns3-server 2.2.0 running on Ubuntu 18.04)

Setting the 'host' setting changes the binding for the main API, but does not change the binding for console port forwarding. e.g.

```
==> ~/.config/GNS3/2.2/gns3_server.conf <==
[Server]
host = 192.168.122.1
```

After restarting gns3:

```
# netstat -natp | grep python
tcp        0      0 0.0.0.0:5001            0.0.0.0:*               LISTEN      1814/python
tcp        0      0 192.168.122.1:3080      0.0.0.0:*               LISTEN      1814/python
tcp        0      0 192.168.122.1:59790     192.168.122.1:43721     ESTABLISHED 1814/python
tcp        0      0 192.168.122.1:42320     192.168.122.1:3080      ESTABLISHED 1814/python
tcp        0      0 192.168.122.1:3080      192.168.122.1:42320     ESTABLISHED 1814/python
tcp        0      0 192.168.122.1:60382     192.168.122.1:38767     ESTABLISHED 1814/python
tcp        0      0 127.0.0.1:34660         127.0.0.1:5002          ESTABLISHED 1814/python
```

Note that the listening socket on port 5001 is still bound to 0.0.0.0.

The actual VM console is 5002 and bound to localhost, as expected:

```
# ps auxwww | grep kvm
... /usr/bin/qemu-system-x86_64 ... -serial telnet:127.0.0.1:5002,server,nowait ...
```

This also prevents consoles working over IPv6.  If I set `host = ::` then the main API on port 3080 is bound to `::`, but the console ports remain bound to IPv4 0.0.0.0 only.

**Comment by candlerb** [comment]
Aside: currently it seems that if you set `host = ::` that the API only accepts connections on v6, not v4.  It would be great if it would accept both (which should be possible on a single socket, if it's suitably initialized)

**Comment by candlerb** [comment]
I think I have found the source of this issue.

To select the address to listen on, GNS3 uses uses the console_host attribute of the PortManager (note: not the console_host attribute of the node!)

```
        self._wrapper_telnet_server = await asyncio.start_server(server.run, self._manager.port_manager.console_host, self.console)
```

In `gns3server/web/web_server.py`, the console_host property of the PortManager is set to be the same as the `host` configuration:

```
        PortManager.instance().console_host = self._host
```

... but in `less gns3server/compute/port_manager.py` it is overridden to 0.0.0.0 if `allow_remote_console` is set.

```
    @console_host.setter
    def console_host(self, new_host):
        """
        Bind console host to 0.0.0.0 if remote connections are allowed.
        """

        server_config = Config.instance().get_section_config("Server")
        remote_console_connections = server_config.getboolean("allow_remote_console")
        if remote_console_connections:
            log.warning("Remote console connections are allowed")
            self._console_host = "0.0.0.0"
        else:
            self._console_host = new_host
```

This has the odd consequence that if `remote_console_connections` is set to false, the value of new_host is honoured (which may allow remote console connections).

I wonder if this should be the other way round: use the supplied value for new_host unless allow_remote_console is not set, in which case force it to 127.0.0.1 ?

But perhaps the intention of the "allow_remote_console" option was to bind consoles to 0.0.0.0 even if the server itself was bound to 127.0.0.1?

In that case I suggest the following:

```
    @console_host.setter
    def console_host(self, new_host):
        """
        Bind console host to supplied address, or wider if remote connections are allowed.
        """

        server_config = Config.instance().get_section_config("Server")
        remote_console_connections = server_config.getboolean("allow_remote_console")
        if not remote_console_connections:
            self._console_host = "127.0.0.1"
        else:
            log.warning("Remote console connections are allowed")
            if new_host == "127.0.0.1":
                self._console_host = "0.0.0.0"
            elif new_host == "::1":
                self._console_host = "::"
            else:
                self._console_host = new_host
```

This works for me - I can format this as a PR if you want to do this.

------

Some other notes from digging around the code.

I was a bit confused by this code in `gns3server/controller/__init__.py`

```
        # clients will use the IP they use to connect to
        # the controller if console_host is 0.0.0.0
        console_host = host
        if host == "0.0.0.0":
            host = "127.0.0.1"   <<< ??
...
            self._local_server = await self.add_compute(compute_id="local",
                                                        name=name,
                                                        protocol=server_config.get("protocol", "http"),
                                                        host=host,
                                                        console_host=console_host,
```

The question is: why would you set `host` to 127.0.0.1, whilst leaving `console_host` at 0.0.0.0?

Based on what I said above, plus some experiments - see https://github.com/GNS3/gns3-web-ui/issues/558 [^1] - I believe `console_host` is the address to connect to, rather than the address to listen on, for console connections.

The above code leaves the console_host as 0.0.0.0 if host is 0.0.0.0.  That's find if it pushes back to the client to work out for itself what address to use. [^1]

But in that case, why set host to 127.0.0.1 ?

----

[^1] The webUI seems to display the `console_host` verbatim, so tells you to `telnet 0.0.0.0:5001`, whilst the regular GUI shows the remote server name or address to telnet to.

**Comment by candlerb** [comment]
There was one other piece to this puzzle.  In my systemd service file, it turns out I was running gns3server with the `-A` option, which overrides the `allow_remote_console` setting.

So to summarise: with the existing code I can get the behaviour I want (which is to bind both the API and console ports to 192.168.122.1) by setting:

```
[Server]
host = 192.168.122.1
allow_remote_console = False
```

(as long as gns3server is not started with `-A` option, which forces `allow_remote_console` to True)

However, if you set

```
[Server]
;host = 0.0.0.0 is the default
allow_remote_console = False
```

then you get the unexpected result that both API and consoles permit remote access - i.e. both are bound to 0.0.0.0!

In other words, `allow_remote_console=yes` really means "force console to bind to 0.0.0.0, regardless of what the API is bound to"; while `allow_remote_console=no` means "bind console access the same as the API" (which may or may not allow remote connections)

Phew!

**Comment by grossmj** [comment]
No worries. Do you think we should rename the `allow_remote_console` setting to something more obvious?

**Comment by candlerb** [comment]
I wondered that too, but I couldn't think of a good alternative. `bind_console_any`? `force_remote_console`?

It loses the naming connection to the `-A / --allow` flag, although `any` helps a bit.

Changing either the config setting or the CLI flag has backwards-compatibility issues, so you might as well change both; keep new and old in parallel until the next major release.

**Comment by grossmj** [comment]
I think I won't touch this due to backwards-compatibility. Thanks :)
